File: tests/op_table.test.js
import { describe, it } from "node:test";
import assert from "node:assert";

import OpTable from "../op_table.js";

describe("test SIC opcode table", () => {
  const table = new OpTable();

  it("should find opcode by valid mnemonic", () => {
    assert.strictEqual(table.find("ADD"), 0x18);
    assert.strictEqual(table.find("ADDF"), 0x58);
    assert.strictEqual(table.find("ADDR"), 0x90);
    assert.strictEqual(table.find("AND"), 0x40);
    assert.strictEqual(table.find("CLEAR"), 0xb4);
    assert.strictEqual(table.find("COMP"), 0x28);
    assert.strictEqual(table.find("COMPF"), 0x88);
    assert.strictEqual(table.find("COMPR"), 0xa0);
    assert.strictEqual(table.find("DIV"), 0x24);
    assert.strictEqual(table.find("DIVF"), 0x64);
    assert.strictEqual(table.find("DIVR"), 0x9c);
    assert.strictEqual(table.find("FIX"), 0xc4);
    assert.strictEqual(table.find("FLOAT"), 0xc0);
    assert.strictEqual(table.find("HIO"), 0xf4);
    assert.strictEqual(table.find("J"), 0x3c);
    assert.strictEqual(table.find("JEQ"), 0x30);
    assert.strictEqual(table.find("JGT"), 0x34);
    assert.strictEqual(table.find("JLT"), 0x38);
    assert.strictEqual(table.find("JSUB"), 0x48);
    assert.strictEqual(table.find("LDA"), 0x00);
    assert.strictEqual(table.find("LDB"), 0x68);
    assert.strictEqual(table.find("LDCH"), 0x50);
    assert.strictEqual(table.find("LDF"), 0x70);
    assert.strictEqual(table.find("LDL"), 0x08);
    assert.strictEqual(table.find("LDS"), 0x6c);
    assert.strictEqual(table.find("LDT"), 0x74);
    assert.strictEqual(table.find("LDX"), 0x04);
    assert.strictEqual(table.find("LPS"), 0xd0);
    assert.strictEqual(table.find("MUL"), 0x20);
    assert.strictEqual(table.find("MULF"), 0x60);
    assert.strictEqual(table.find("MULR"), 0x98);
    assert.strictEqual(table.find("NORM"), 0xc8);
    assert.strictEqual(table.find("OR"), 0x44);
    assert.strictEqual(table.find("RD"), 0xd8);
    assert.strictEqual(table.find("RMO"), 0xac);
    assert.strictEqual(table.find("RSUB"), 0x4c);
    assert.strictEqual(table.find("SHIFTL"), 0xa4);
    assert.strictEqual(table.find("SHIFTR"), 0xa8);
    assert.strictEqual(table.find("SIO"), 0xf0);
    assert.strictEqual(table.find("SSK"), 0xec);
    assert.strictEqual(table.find("STA"), 0x0c);
    assert.strictEqual(table.find("STB"), 0x78);
    assert.strictEqual(table.find("STCH"), 0x54);
    assert.strictEqual(table.find("STF"), 0x80);
    assert.strictEqual(table.find("STI"), 0xd4);
    assert.strictEqual(table.find("STL"), 0x14);
    assert.strictEqual(table.find("STS"), 0x7c);
    assert.strictEqual(table.find("STSW"), 0xe8);
    assert.strictEqual(table.find("STT"), 0x84);
    assert.strictEqual(table.find("STX"), 0x10);
    assert.strictEqual(table.find("SUB"), 0x1c);
    assert.strictEqual(table.find("SUBF"), 0x5c);
    assert.strictEqual(table.find("SUBR"), 0x94);
    assert.strictEqual(table.find("SVC"), 0xb0);
    assert.strictEqual(table.find("TD"), 0xe0);
    assert.strictEqual(table.find("TIO"), 0xf8);
    assert.strictEqual(table.find("TIX"), 0x2c);
    assert.strictEqual(table.find("WD"), 0xdc);
  });

  it("should return undefined for invalid mnemonic", () => {
    assert.strictEqual(table.find("UNKNOWN"), undefined);
  });
});

File: op_table_test.go
package optable

import "testing"

func TestFindOpcodeByValidMnemonic(t *testing.T) {
	tests := []struct {
		mnemonic       string
		expectedOpcode uint8
	}{
		{"ADD", 0x18},
		{"ADDF", 0x58},
		{"ADDR", 0x90},
		{"AND", 0x40},
		{"CLEAR", 0xB4},
		{"COMP", 0x28},
		{"COMPF", 0x88},
		{"COMPR", 0xA0},
		{"DIV", 0x24},
		{"DIVF", 0x64},
		{"DIVR", 0x9C},
		{"FIX", 0xC4},
		{"FLOAT", 0xC0},
		{"HIO", 0xF4},
		{"J", 0x3C},
		{"JEQ", 0x30},
		{"JGT", 0x34},
		{"JLT", 0x38},
		{"JSUB", 0x48},
		{"LDA", 0x00},
		{"LDB", 0x68},
		{"LDCH", 0x50},
		{"LDF", 0x70},
		{"LDL", 0x08},
		{"LDS", 0x6C},
		{"LDT", 0x74},
		{"LDX", 0x04},
		{"LPS", 0xD0},
		{"MUL", 0x20},
		{"MULF", 0x60},
		{"MULR", 0x98},
		{"NORM", 0xC8},
		{"OR", 0x44},
		{"RD", 0xD8},
		{"RMO", 0xAC},
		{"RSUB", 0x4C},
		{"SHIFTL", 0xA4},
		{"SHIFTR", 0xA8},
		{"SIO", 0xF0},
		{"SSK", 0xEC},
		{"STA", 0x0C},
		{"STB", 0x78},
		{"STCH", 0x54},
		{"STF", 0x80},
		{"STI", 0xD4},
		{"STL", 0x14},
		{"STS", 0x7C},
		{"STSW", 0xE8},
		{"STT", 0x84},
		{"STX", 0x10},
		{"SUB", 0x1C},
		{"SUBF", 0x5C},
		{"SUBR", 0x94},
		{"SVC", 0xB0},
		{"TD", 0xE0},
		{"TIO", 0xF8},
		{"TIX", 0x2C},
		{"TIXR", 0xB8},
		{"WD", 0xDC},
	}

	table := NewOpTable()
	for _, tt := range tests {
		opcode, ok := table.Find(tt.mnemonic)
		if ok != true || opcode != tt.expectedOpcode {
			t.Errorf("Find(%s) = %d, %t; want %d",
				tt.mnemonic, opcode, ok, tt.expectedOpcode)
		}
	}
}

func TestFindOpcodeByInvalidMnemonic(t *testing.T) {
	table := NewOpTable()
	_, ok := table.Find("UNKNOWN")
	if ok != false {
		t.Errorf("Find(UNKNOWN) = _, %t; want false", ok)
	}
}
